Chore/36 authenticate customers with backend apis

DavJesse opened the pull request:
## Summary of Changes
### 1. Core Authentication Logic
Introduced two custom React hooks to decouple business logic from the `AuthPage` component:

- `useLoginCustomer`: Manages the login lifecycle, including state for `isLoading` and server-side `error` handling.
- `useRegisterCustomer`: Handles complex customer registration, mapping UI state to backend requirements and managing multi-status error handling (e.g., 409 Conflict, 429 Rate Limiting).
  - **Fix:** Implemented a robust check for 429 status codes to handle non-JSON error responses (plain text) gracefully without crashing the application.

### 2. Enhanced Registration Security
Expanded the registration form to include standard security practices:
- Added **Password** and **Confirm Password** fields.
- Implemented client-side validation logic to ensure passwords match and meet a minimum length of **8 characters.**
- Added a "Missing Field" validator that identifies empty inputs and provides human-readable error messages (e.g., "Zip Code is required") instead of technical keys.

### 3. Smart Input Component Upgrades
Refactored the `InputField` component to be context-aware:

- **Auto-Detection:** The component now detects when `type="password"` is passed.
- **Visibility Toggle:** Integrated a dynamic "Show/Hide" toggle using `lucide-react` icons. This improves UX by allowing users to verify their password entry before submission.
- **Focus States:** Added improved CSS transitions for icons and borders to reflect active focus and error states.

### 4. UI/UX Improvements
- **Paginated Registration:** Divided the registration form into two logical steps (Personal Details → Address & Profile) to reduce cognitive load.
- **Dynamic Metadata:** Integrated `updateMeta` to update page titles and SEO descriptions dynamically based on whether the user is on the Login or Register view.
- **Centralized Error Display:** Replaced disruptive `alert()` boxes with an inline, animated error area that displays both client-side validation errors and server-side responses.

## How to Test
1. Navigate to `/shop/auth`.
2. **Validation:** Try to submit the registration form with an empty field. You should see a red error message identifying the specific field.
3. **Password Check:** Enter mismatched passwords and verify the error message appears in the UI.
4. **Visibility:** Click the eye icon in the password field to toggle text visibility.
5. **Rate Limiting:** (Optional) Trigger a 429 response from the backend and verify that the message "Too many requests. Please try again later." is displayed correctly.

**Related Issues:** closes #36

DavJesse committed:
chore: parse customer login data (#36)

DavJesse committed:
chore: update gitignore (#36)

DavJesse committed:
feat: add useLoginCustomer hook to handle login customer (#36)

DavJesse committed:
chore: add .env.example (#36)

DavJesse committed:
feat: add custom hook to register customers (#36)

DavJesse committed:
chore: implement useRegisterCustomer to register customers (#36).
Tweak regitration and login forms UI (#36)

DavJesse committed:
Merge branch 'main' into chore/36-authenticate-customers-with-backend-apis

Murzuqisah committed:
Implement Google OAuth authentication

- Added google-auth-library dependency
- Extended user schema with google_id, avatar_url, auth_provider fields
- Made password field nullable for OAuth users
- Added Google token verification in AuthService
- Implemented Google user creation and lookup methods
- Added /api/auth/google endpoint for Google login
- Enhanced user repository with findByGoogleId method
- Added Google OAuth environment variables
- Fixed TypeScript errors for nullable password field

DavJesse committed:
chore: update backend dependancies (#36)

DavJesse committed:
chore: update frontend dependancies (#36)

DavJesse committed:
fix: introduce error state to track mismatched and short passwords (#36)

DavJesse committed:
fix: check error status before parsing json (#36)

DavJesse committed:
fix: restore router to recognize correct routes (#36)

nyunja reviewed (changes requested):
Your branch has this database issue

<img width="1443" height="754" alt="Screenshot from 2025-12-14 13-23-06" src="https://github.com/user-attachments/assets/5bef71ea-1535-4c66-96f1-dbcce6cb8a77" />

Murzuqisah committed:
Fix Google OAuth authentication errors

- Fixed createUser method to include Google OAuth fields with defaults
- Added proper Google ID extraction from JWT payload (payload.sub)
- Fixed token verification to return google_id in response
- Added validation for payload.sub existence
- Resolved 500 errors in both local and Google OAuth registration
- Enhanced error handling for missing Google tokens

DavJesse commented:
I hope @Murzuqisah explained to you how to get rid of this error. It is a clash between your outdated local DB and the updated schema

Murzuqisah committed:
fix: update start.sh with correct ports and WSL compatibility

Murzuqisah committed:
feat: add frontend Google OAuth integration with redirect flow

DavJesse committed:
Merge branch 'feature/implement-google-authentication-support' into chore/36-authenticate-customers-with-backend-apis

DavJesse commented:
I have merged @Murzuqisah 's `feature/implement-google-authentication-support` into this branch. That should take care of the migration issues.

A few notes when setting up:
- Delete you DB if you have any
- Run `nmp install` on both `root` and `/backend`, to ensure packages are up to date
- Run `./start.sh` to start both backend and frontend servers
- If you database still lacks tables, navigate to `/backend` then run `npm run migrate`
- Test authentication on `/shop/auth` registation and login should work, I will implement Google on a later PR
